### DISCUSSION TITLE
Use GCP Batch or Cloud Run Jobs for Runs?

### DISCUSSION THREAD

**Discussion by seandavi** [dagster-io/dagster#18958] [Ideas]
Google Cloud has a couple of batch processing environments:

- [GCP Batch](https://cloud.google.com/batch)
- [Cloud Run Jobs](https://cloud.google.com/run/docs/create-jobs)

Is anyone working on a Dagster integration for running jobs using either of these services, akin to the [DockerRunLauncher](python_modules/libraries/dagster-docker/dagster_docker/docker_run_launcher.py)?

**Comment by tim-carbonfarm** [comment]
I have set up a "GCP-native" Dagster deployment which:
- uses Cloud Run (services) for gRPC code location servers
- uses Cloud Run Jobs as run workers

So far it's working very well. If there is some interest I could look into open-sourcing it.

**Reply by mudravrik** [reply]
@tim-carbonfarm hi!
As a Dagster-on-GCP user, would love to take a look on it!

**Reply by aldomann** [reply]
This sounds really interesting, as a Dagster and Cloud Run user, being able to use both in unison would be great. Do you have any example or code to replicate your set up?

**Reply by indigoviolet** [reply]
@tim-carbonfarm yes - we would be very interested in this as well!

**Reply by timchap** [reply]
[Here's the PR](https://github.com/dagster-io/dagster/pull/21864) for the run launcher implementation. As mentioned, I will also try to share some Terraform code when I get the opportunity.

**Reply by garethbrickman** [reply]
@timchap Making you aware of this earlier PR: https://github.com/dagster-io/dagster/pull/21509

**Reply by tim-carbonfarm** [reply]
Thanks for the heads up, looks like I was just too late :) happy to close my PR or otherwise combine it with the pre-existing one as you see fit

**Reply by garethbrickman** [reply]
@baumann-t Looping you in to this discussion

**Reply by baumann-t** [reply]
Thank you @garethbrickman for making me aware of this. Happy to work with you and @tim-carbonfarm / @timchap to see how we can move forward with the best Cloud Run Launcher!

**Reply by timchap** [reply]
Sorry, should really merge my GH accounts :D feel free to ping me on either.

The main implementation difference I see is that #21509 does not currently include support for multiple code locations. This can be added in one of several ways:
- as I have done (with a mapping from code location name -> job name)
- per my first implementation, by specifying `job_name_prefix` in the run launcher config. With this approach, we will look for a Cloud Run job with the name `job_name_prefix + code_location_name`. I moved away from this approach because I found it a bit opaque.
- or finally, by dynamically creating the Job resource for each run. I personally don't like this approach because I'm pretty deep in the declarative infrastructure game, and it seems a bit contrary to GCP's intended use of Cloud Run Jobs. But this approach would be more analogous to the EcsRunLauncher.

Perhaps there are other options, I'm open to suggestions. Otherwise, the main things to consolidate are:
- Unit tests (currently missing from #21509)
- Examples (currently missing from #21864)
- Integration tests if desired (open to your suggestions here @garethbrickman)
- Terraform code if desired (ditto)

**Reply by baumann-t** [reply]
@timchap 
For the management of multiple code locations: I like your current implementation (with a mapping from code location name -> job name). As you said, it seems more declarative and less opaque than using a naming convention. Agree also on not dynamically creating a Job resource per run.

For consolidation, I would suggest:
- Let's keep your Launcher as it already includes the code location management and the unit tests
- We can add my Examples to make it easier for users (and reviewers) to try the launcher. We could either adapt my script to deploy multiple Run Job Launchers to test the feature, or add some Terraform code; but I don't think Terraform is really appropriate for Examples in Dagster (as you said, @garethbrickman can guide us on this). 

Let me know what you think and we how you want to proceed with that!

**Reply by garethbrickman** [reply]
Thanks so much folks! @baumann-t would you please go ahead and close your PR and move over what's needed into https://github.com/dagster-io/dagster/pull/21864. I'll then get input from the Dagster engineering team about integration tests and next steps! 👍

**Reply by baumann-t** [reply]
Perfect! PR #21509 is closed. I received an invitation from @timchap to contribute to his PR. I've copied my existing Example and pushed it to his repo. I should have more time later this week to adapt my original example to the new launcher.

@timchap, I think we could add both of our approaches for the gRPC code location servers in the Cloud Run Launcher Example:
- One example using a simple VM (like mine)
- One example using Cloud Run Services (like what you mentioned)

This would showcase two different options for running Dagster on GCP. Let me know what you think!

**Reply by leonardovff** [reply]
the pr was completely ignored :\

**Reply by garethbrickman** [reply]
Hey folks, can assure you the PR has not been ignored. Please check the latest reply here: https://github.com/dagster-io/dagster/pull/21864#issuecomment-2359242800

**Reply by garethbrickman** [reply]
Work on the GCP Cloud Runner has been migrated to our community-integrations repo: https://github.com/dagster-io/community-integrations/pull/26

**Reply by garethbrickman** [reply]
The GCP Cloud Runner is now available in its community-integrations repo: [dagster-contrib-gcp]( https://github.com/dagster-io/community-integrations/tree/main/libraries/dagster-contrib-gcp)

And can be installed by running: 
```
pip install dagster_contrib_gcp
```
